src/core/server/core_app/bundle_routes/register_bundle_routes.ts

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { join } from 'path';
+import { PackageInfo } from '@kbn/config';
+import { distDir as uiSharedDepsDistDir } from '@kbn/ui-shared-deps';
+import { IRouter } from '../../http';
+import { UiPlugins } from '../../plugins';
+import { fromRoot } from '../../utils';
+import { FileHashCache } from './file_hash_cache';
+import { registerRouteForBundle } from './bundles_route';
+
+/**
+ *  Creates the routes that serves files from `bundlesPath`.
+ *
+ *  @param {Object} options
+ *  @property {Array<{id,path}>} options.npUiPluginPublicDirs array of ids and paths that should be served for new platform plugins
+ *  @property {string} options.regularBundlesPath
+ *  @property {string} options.basePublicPath
+ *
+ *  @return Array.of({Hapi.Route})
+ */
+export function registerBundleRoutes({
+  router,
+  serverBasePath, // serverBasePath
+  uiPlugins,
+  packageInfo,
+}: {
+  router: IRouter;
+  serverBasePath: string;
+  uiPlugins: UiPlugins;
+  packageInfo: PackageInfo;
+}) {
+  const { dist: isDist, buildNum } = packageInfo;
+  // rather than calculate the fileHash on every request, we
+  // provide a cache object to `resolveDynamicAssetResponse()` that
+  // will store the 100 most recently used hashes.
+  const fileHashCache = new FileHashCache();
+
+  registerRouteForBundle(router, {
+    publicPath: `${serverBasePath}/${buildNum}/bundles/kbn-ui-shared-deps/`,
+    routePath: `/${buildNum}/bundles/kbn-ui-shared-deps/`,
+    bundlesPath: uiSharedDepsDistDir,
+    fileHashCache,
+    isDist,
+  });
+  registerRouteForBundle(router, {
+    publicPath: `${serverBasePath}/${buildNum}/bundles/core/`,
+    routePath: `/${buildNum}/bundles/core/`,
+    bundlesPath: fromRoot(join('src', 'core', 'target', 'public')),
+    fileHashCache,
+    isDist,
+  });
+
+  [...uiPlugins.internal.entries()].forEach(([id, { publicTargetDir }]) => {
+    registerRouteForBundle(router, {
+      publicPath: `${serverBasePath}/${buildNum}/bundles/plugin/${id}/`,
+      routePath: `/${buildNum}/bundles/plugin/${id}/`,
+      bundlesPath: publicTargetDir,
+      fileHashCache,
+      isDist,
+    });
+  });
```

Comment:
We could probably optimize this a little to have a single route registered to serve all plugins bundle. I didn't bother, as it created divergence in the underlying `registerRouteForBundle` between plugins and non-plugins bundles, and just introduced more complexity, which isn't worth it ihmo.